src/core/ddsi/src/q_xmsg.c

```diff
@@ -573,6 +578,11 @@ void nn_xmsg_submsg_replace(struct nn_xmsg *msg, struct nn_xmsg_marker sm_marker
 
   /* Replace the sub-message. */
   memcpy(msg->data->payload + sm_marker.offset, new_submsg, new_len);
+
+  /* if it is a DATA_REXMIT one, reset the readerId offset because it may no longer be valid,
+     or patching it might break stuff if the replaced data is encrypted or authenticated */
```

Comment:
This comment is not completely clear to me. I see that `lookup_msg` doesn't return the msg when it is of kind `NN_XMSG_KIND_DATA_REXMIT_NOMERGE` and therefore security-encoded messages won't be merged. But maybe you could add in this comment at what point 'the readerId offset' is reset?

Reply:
Ouch. I changed my mind: first I thought I'd reset the readerId and use that as an indication, then later decided to use a different kind and failed to update the comment.